compiler-core/src/type_/expression.rs

```diff
@@ -3363,20 +3371,107 @@ impl<'a, 'b> ExprTyper<'a, 'b> {
                         value: record_access,
                         implicit: Some(ImplicitCallArgOrigin::RecordUpdate),
                     },
                 ))
-            })
-            .collect::<Result<Vec<_>, _>>()?;
+            } else {
+                let (accessor_type, positional_fields) =
+                    match collapse_links(record.type_()).as_ref() {
+                        // A type in the current module
+                        Type::Named {
+                            module,
+                            name,
+                            inferred_variant,
+                            ..
+                        } if module == &self.environment.current_module => {
+                            self.environment
+                                .accessors
+                                .get(name)
+                                .and_then(|accessors_map| {
+                                    Some((
+                                        accessors_map.type_.clone(),
+                                        // For record updates we must know the variant of the record, so if the
+                                        // variant has not been inferred, that means there must only be one.
+                                        accessors_map
+                                            .positional_accessors(inferred_variant.unwrap_or(0))?,
+                                    ))
+                                })
+                        }
+
+                        // A type in another module
+                        Type::Named {
+                            module,
+                            name,
+                            inferred_variant,
+                            ..
+                        } => self
+                            .environment
+                            .importable_modules
+                            .get(module)
+                            .and_then(|module| module.accessors.get(name))
+                            .filter(|a| {
+                                a.publicity.is_importable()
+                                    || module == &self.environment.current_module
+                            })
+                            .and_then(|accessors_map| {
+                                Some((
+                                    accessors_map.type_.clone(),
+                                    accessors_map
+                                        .positional_accessors(inferred_variant.unwrap_or(0))?,
+                                ))
+                            }),
+
+                        _ => panic!("Type has already checked to be valid"),
+                    }
+                    .expect("Variant has already checked to be valid");
+
+                let type_ = positional_fields
+                    .get(index as usize)
+                    .expect("Field exists")
+                    .clone();
+                let mut type_vars = im::HashMap::new();
```

Comment:
Won't this result in rigid type variables not being respected? Without them being passed in they would be instantiated to an unbound variable instead of remaining generic?